src/tools/singlejar/log4j2_plugin_dat_combiner.cc

```diff
@@ -0,0 +1,154 @@
+#include "src/tools/singlejar/log4j2_plugin_dat_combiner.h"
+
+#include <sstream>
+#include <string>
+#include <vector>
+
+template<typename T>
+inline static T swapByteOrder(const T& val) {
+  int totalBytes = sizeof(val);
+  T swapped = (T) 0;
+  for (int i = 0; i < totalBytes; ++i) {
+    swapped |= (val >> (8*(totalBytes-i-1)) & 0xFF) << (8*i);
+  }
+  return swapped;
+}
+
+bool readBool(std::istringstream &stream) {
+  bool value;
+  stream.read(reinterpret_cast<char *>(&value), sizeof(value));
+  return value;
+}
+
+uint32_t readInt(std::istringstream &stream) {
+  uint32_t values;
+  stream.read(reinterpret_cast<char *>(&values), sizeof(values));
+  return swapByteOrder(values);
+}
+
+std::string readUTFString(std::istringstream &stream) {
+  uint16_t length;
+  stream.read(reinterpret_cast<char *>(&length), sizeof(length));
+  length = swapByteOrder(length); // Convert to host byte order
```

Comment:
Since I totally lack context here - and a future code maintainer may do so, too: Could you add a sentence or two more as to why the byte order needs to be swapped? And maybe the best place for that comment would be at the swapByteOrder() function itself?
I'm guessing it is because JVM class files are big-endian, and x86 and most ARMs are little-endian?
I guess there is not much big-endian hardware around these days, but I may be wrong, it's a long time since last I encountered endianness as a topic.

Reply:
I added a comment to that effect in https://github.com/bazelbuild/bazel/pull/22581/commits/c80889e8e7cc2bb9e640b56b0803a814103fa9ec.